File: 0005.LongestPalindromicSubstring.go
// https://leetcode.com/submissions/detail/77520312/ solution is better
func longestPalindrome(s string) string {
    n := len(s)
    idx, maxlen := 0, 0
    for i := 0; i < n; i++ {
        if n - i <= maxlen>>1 {
            break
        }
        expand(s, i, i, &idx, &maxlen)
        expand(s, i, i+1, &idx, &maxlen)
    }
    return s[idx:idx+maxlen]
}

func expand(s string, i, j int, idx, maxlen *int) {
    for ; i >= 0 && j < len(s) && s[i] == s[j]; i, j = i-1, j+1 {}
    if j-i-1 > *maxlen {
        *idx = i+1
        *maxlen = j-i-1
    }
} 

File: 0005.LongestPalindromicSubstring.py
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        length = len(s)
        if length <= 1:
            return s
        base, maxLen = 0, 1
        i = 0
        while i < length:
            if length - base <= maxLen/2:
                break
            l = r = i
            while r < length-1 and s[r+1] == s[r]:
                r += 1
            i = r + 1
            while l > 0 and r < length-1 and s[l-1] == s[r+1]:
                l -= 1
                r += 1
            if r-l+1 > maxLen:
                base, maxLen = l, r-l+1
        return s[base:base+maxLen]

print Solution().longestPalindrome('dfgggfk')